bin/standupbot.py

```diff
@@ -2,205 +2,253 @@
 
 from __future__ import print_function
 
-import re
+import argparse
 import json
-import sys
+import logging
+import os
 import random
+import re
+import sys
 from datetime import datetime, timedelta
 
-from future.moves.urllib.request import urlopen, Request
 from future.moves.urllib.error import HTTPError
-
-from mediawiki import MediaWiki, CONFIG
-
-import argparse
-import logging
+from future.moves.urllib.request import Request, urlopen
+from mediawiki import CONFIG, MediaWiki
 
 logger = logging.getLogger('wikibot')
 logging.basicConfig()
 
-parser = argparse.ArgumentParser()
-parser.add_argument("-v", "--verbose", help="increase output verbosity", action="store_true")
-parser.add_argument("-q", "--quiet", help="don't actually message Slack", action="store_true")
-parser.add_argument("-a", "--all", help="announce all standups, regardless of time", action="store_true")
-parser.add_argument("-d", "--date", help="provide fake date to use")
-parser.add_argument("-c", "--channel", help="announce in a specific channel, such as @username or #sandbox")
-args = parser.parse_args()
 
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-v", "--verbose", help="increase output verbosity", action="store_true")
+    parser.add_argument("-q", "--quiet", help="don't actually message Slack", action="store_true")
+    parser.add_argument("-a", "--all", help="announce all standups, regardless of time", action="store_true")
+    parser.add_argument("-d", "--date", help="provide fake date to use")
+    parser.add_argument("-c", "--channel", help="announce in a specific channel, such as @username or #sandbox")
+    args = parser.parse_args()
+    return args
+
+
+def no_run_date(time):
+    if time.weekday() >= 5:
+        logger.info('Today is the weekend, we don\'t have any standups.')
+        return True
+
+    path = os.path.dirname(__file__) + '/../../bank-holidays.json'
+    try:
+        data = json.load(open(path))
+    except IOError:
+        return False
+
+    date = time.date().isoformat()
+    def is_today(i):
+        return i['date'] == date
+    bh_ew = filter(is_today, data['england-and-wales']['events'])
+    bh_s = filter(is_today, data['scotland']['events'])
+
+    if bh_ew and bh_s:
+        logger.info('UK-wide bank holiday, no standups')
+        return True
+
+    return False
+
+
+class WikiParser(object):
+    # Regex patterns
+    listPattern = re.compile(r'\* (.*)')
+    planningPattern = re.compile('<!-- Begin Planning Schedule -->(.*?)<!-- End Planning Schedule -->', re.S)
+    sectionPattern = re.compile('<!-- Begin Standups Schedule -->(.*?)<!-- End Standups Schedule -->', re.S)
+    entryPattern = re.compile('\\|-\n\\| (.*?)\n\\| (.*?)\n\\| (.*?)\n\\| (.*?)\n\\| (.*?)\n\\| (.*?)\n', re.S)
```

Comment:
I'm curious, why add the additional backslashes, given this already works? Is it because this will be deprecated in future under Python 3?

I see we can't use `r` because of the need to match `\n` and I'm aware of the notes on the use of `\` in the docs (which also include the phrase "This is complicated and hard to understand"), but the notes on `|` say  "To match a literal `|`, use `\|`"- hence my question as I am a bit confused!

Reply:
Yeah, it throws a warning now (since 3.6) and will become an error at some point. The explanation right at the top of https://docs.python.org/3/library/re.html is probably easier to understand than the bit you quote. In a non-raw string, the backslash gets looked at by the string parser first - so `\\` will become `\` at that point - and then the regex parser, so it's still `\|` for that (and that's the second bit of docs you quote). If it were just `\|` then it does get through the python parser at present, which leaves it alone, but at some point it will not. Could use a raw string and concatenate in the newlines, I guess.